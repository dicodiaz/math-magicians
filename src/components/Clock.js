import PropTypes from 'prop-types';
import React from 'react';

export default class Clock extends React.Component {
  constructor(props) {
    super(props);
    this.state = { date: new Date() };
  }

  componentDidMount() {
    const { updateEveryNSeconds } = this.props;
    this.timerID = setInterval(() => this.tick(), 1000 * updateEveryNSeconds);
  }

  componentWillUnmount() {
    clearInterval(this.timerID);
  }

  tick() {
    this.setState({
      date: new Date(),
    });
  }

  render() {
    const { date } = this.state;
    return (
      <div className="col-md-9">
        <div className="row text-end text-white">
          <p>{date.toLocaleTimeString()}</p>
        </div>
      </div>
    );
  }
}

Clock.propTypes = {
  updateEveryNSeconds: PropTypes.number.isRequired,
};
