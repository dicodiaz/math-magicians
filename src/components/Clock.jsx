import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';

const Clock = ({ updateEveryNSeconds }) => {
  const [date, setDate] = useState(new Date());

  useEffect(() => {
    const timerID = setInterval(() => setDate(new Date()), 1000 * updateEveryNSeconds);
    return () => clearInterval(timerID);
  }, [updateEveryNSeconds]);

  return (
    <div className="col-md-9">
      <div className="row text-end text-white">
        <p data-testid="time">{date.toLocaleTimeString()}</p>
      </div>
    </div>
  );
};

Clock.propTypes = {
  updateEveryNSeconds: PropTypes.number.isRequired,
};

export default Clock;
