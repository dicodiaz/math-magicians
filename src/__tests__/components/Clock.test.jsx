import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import Clock from '../../components/Clock';

describe('Clock component', () => {
  const mockDate = new Date('2023-03-05T00:00:00');
  const spy = jest.spyOn(global, 'Date').mockImplementationOnce(() => mockDate);

  afterAll(() => {
    spy.mockRestore();
  });

  it('should render as expected', async () => {
    const { container } = render(<Clock updateEveryNSeconds={1} />);
    expect(container).toMatchSnapshot();
  });
});
