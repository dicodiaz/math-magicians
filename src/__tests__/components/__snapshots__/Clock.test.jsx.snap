// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clock component should render as expected 1`] = `
<div>
  <div
    class="col-md-9"
  >
    <div
      class="row text-end text-white"
    >
      <p
        data-testid="time"
      >
        12:00:00 a. m.
      </p>
    </div>
  </div>
</div>
`;
